### PR TITLE
Use data provider to test both api versions

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Adds a compatibility layer and converts some api3 tests to also run on api v4.

Before
----------------------------------------
Core api tests covered v3 only.

After
----------------------------------------
311 v3 tests converted to also test v4.

Technical Details
----------------------------------------
Uses `@dataProvider` annotation to feed 2 version numbers to each test.

Comments
----------------------------------------
The key issue with apiv4 is that it doesn't do enough - ie it does the metadata crud stuff but not enough of the other stuff that makes apiv3 useful - all the different actions. 

In order to add the actions to apiv4 we need to figure out how to make the tests test apiv4  -since the tests, rather than the code, are the main asset and we shouldn't even be adding actions without test parity to api v4.

One idea is that we run all our apiv3 tests using apiv4 as well - with some obvious opt outs - goodbye ConstantTest....

This would cause quite a few fails but we can work through them & decide which represent missing v4 features and which should not be ported.

If we do this the question becomes how we mark which tests should run on both versions - ideally we would follow an opt out - ie. by manually doing this or some other method add the dataprovider to all tests & then edit it back with a comment as to why for all the ones we don't port. e.g 'no v4 test as this function is deprecated in v3 & deliberately not ported' or 'no v4 test as we have written a separate test for v4'

### CODE REVIEW

**Comment by colemanw** [comment]
The fact that it fails on line 2 of the test (line 1 being "set api to v4") kind of says it all.
The problem with trying to call both api versions from the same test is that:

1. The inputs are different (select, where, orderBy, etc. instead of flat $params).
2. The actions are different (update instead of create, get+index instead of getsingle, getFields+options instead of getoptions).
3. The outputs are different (slightly, this is the most solvable of the issues IMO).

Which means basically rewriting the test either way.

**Comment by eileenmcnaughton** [comment]
@colemanw actually it fails on line 2 of the test because that is the first api call & in fact apiv4 isn't enabled :-)

However, assuming the failure WAS because getoptions is missing - I would argue that is exactly the point of running things like this.

Assuming the goal of api v4 is to eventually replace apiv3 such than in about 4 years time we can rip apiv3 out of core we need to make it something that people will want to and be able to adopt.  Ideally over the next 6 months we would get to the point where early adopters are using apiv4 for new extensions and then following 6 months we'd aim to get over 50% of devs using apiv4 for new extensions. Year 2 we'd hope to see existing extensions start to convert, year 3 core & year 4 we'd be waiting for the laggards. At the moment getting early adopters to adopt v4 within the next 6 months feels pretty ambitious to me & we'd need to work to make that happen. 

We will need some sort of migration guide and also to make the things easy that we can make easy.

In this case we are saying 'getoptions' is not longer an action - there needs to be a section in the migration guide as to what a converted getoptions call looks like in php, drush, js & ? smarty? Ditto other generic options that we have removed.

Once we've done that we can easily handle getoptions in the callAPISuccess wrapper & expect our users to re-write the code. Or if we thing people getting people to use apiv4 is more important than purity we could handle generic actions like getoptions, getcount in the civicrm_api wrapper so people can focus their code converstion efforts on the tricky stuff rather than write their own wrappers (which is obviously what I would do in my own extensions  if I were converting to use apiv4 since I'd be aiming to reduce the work & risk involved)

@totten

**Comment by eileenmcnaughton** [comment]
closing as this is a discussion piece & doesn't need to be open for that

**Comment by colemanw** [comment]
@eileenmcnaughton I've added some code to translate the v3 syntax for v4 and then convert v4 output back to what v3 tests are expecting. Got that first test passing :)

**Comment by eileenmcnaughton** [comment]
@colemanw YAY. Should I re-open this to see how far it gets or does still stall on apiv4 not being enabled by default (something we should change)

File: tests/phpunit/CiviTest/CiviUnitTestCase.php [hunk 1]
```diff
@@ -1357,12 +1359,6 @@ public function customGroupCreate($params = array()) {
 
     $params = array_merge($defaults, $params);
 
-    //have a crack @ deleting it first in the hope this will prevent derailing our tests
-    $this->callAPISuccess('custom_group', 'get', array(
-      'title' => $params['title'],
-      array('api.custom_group.delete' => 1),
-    ));
```
Comment (colemanw) [inline tests/phpunit/CiviTest/CiviUnitTestCase.php]: @eileenmcnaughton this was breaking the v4 test and on closer look appears to be broken. AFAIK this is incorrect chaining syntax (wrapping the chain param in `array()` shouldn't work) and therefore was not doing anything.

Comment (eileenmcnaughton) [reply tests/phpunit/CiviTest/CiviUnitTestCase.php]: @colemanw there are some weird & wonderful syntaxes that wok but it's easy to test - just do a PR removing just these lines against master

**Comment by colemanw** [comment]
I've reopened to see what happens...

**Comment by seamuslee001** [comment]
Jenkins re test this please

File: tests/phpunit/CiviTest/CiviUnitTestCase.php (hunk 1 above)
Comment (colemanw) [reply tests/phpunit/CiviTest/CiviUnitTestCase.php]: Ok see https://github.com/civicrm/civicrm-core/pull/14239

**Comment by colemanw** [comment]
I'm not sure what to make of the console output - it's not clear to me why this is causing massive test failures but I assume it has something to do with installing the api4 extension.

**Comment by eileenmcnaughton** [comment]
@colemanw - right I don't think the extension is installed/enabled

File: tests/phpunit/CiviTest/CiviUnitTestCase.php [hunk 2]
```diff
@@ -321,6 +321,8 @@ protected function setUp() {
     // disable any left-over test extensions
     CRM_Core_DAO::executeQuery('DELETE FROM civicrm_extension WHERE full_name LIKE "test.%"');
 
+    \CRM_Extension_System::singleton()->getManager()->enable(['org.civicrm.api4']);
+
```
Comment (colemanw) [inline tests/phpunit/CiviTest/CiviUnitTestCase.php]: @eileenmcnaughton this is where this PR enables the extension. It should already be present and from what I can tell jenkins isn't failing to install it.

**Comment by eileenmcnaughton** [comment]
OK - I guess this was the line it fell over on Fatal error: Call to a member function clearMessages() on null in /home/jenkins/bknix-dfl/build/core-14153-1ir8e/sites/all/modules/civicrm/tests/phpunit/CRM/Core/BAO/ActionScheduleTest.php on line 697

**Comment by colemanw** [comment]
@eileenmcnaughton I've gotten quite a few tests to pass locally now, using this pattern. Including tests that use chained api calls + a variety of entities and actions. Yay!
But we still need to figure out why Jenkins hates it so much.

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by colemanw** [comment]
I note with satisfaction that the hybrid api3/4 tests *are actually passing* if you look in the console. But the test run still crashes halfway through for unknown reasons which were at least isolated in #14241

**Comment by colemanw** [comment]
@eileenmcnaughton @seamuslee001 :eyes:!!

**Comment by colemanw** [comment]
Some strange test failure that appear related to transactions and the ids assigned to api requests.

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@colemanw woot - it's passing. In terms of getting it cleaned up for merge maybe 3 commits? One for the general cleanup, one to support v4 tests & one that does the individual tests?

I have some doubt as to whether the dataprovider will be the right method - as opposed to being able to have some way of running through with each version - although there is no harm merging that approach & changing it later. We have a lot of flexibility with tests I think

**Comment by colemanw** [comment]
@eileenmcnaughton yes I want to clean up the commits and add a few more tests. Will have that done soon.

**Comment by colemanw** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
this looks like a mergeable point to me - more can be in follow up prs

**Comment by colemanw** [comment]
@eileenmcnaughton that's 311 tests converted!

**Comment by eileenmcnaughton** [comment]
@colemanw well done - I guess going through individually is the only way though?

**Comment by colemanw** [comment]
@eileenmcnaughton I used a regex replace to add the dataprovider annotation to test functions in bulk, then ran the tests and went through the failures to decide whether it was
a) a problem with the test
b) a problem with the compatibility layer
c) a bug in api4
d) a missing feature in api4
In the process, I fixed quite a few api4 bugs and filled in some missing features, and got the compatibility layer working quite well, but also left a lot in the FIXME pile.

**Comment by eileenmcnaughton** [comment]
@colemanw cool - did you try all functions or did you filter in some cruddy-way?

**Pull request merged**